pkg/ccl/backupccl/alter_backup_test.go

```diff
@@ -0,0 +1,167 @@
+
+	const userfile = "'userfile:///a'"
+	const numAccounts = 1
+
+	tc, sqlDB, _, cleanupFn := backupRestoreTestSetup(t, singleNode, numAccounts, InitManualReplication)
+	execCfg := tc.Server(0).ExecutorConfig().(sql.ExecutorConfig)
+	defer cleanupFn()
+	query := fmt.Sprintf("BACKUP TABLE bank INTO %s WITH kms = %s", userfile, key1)
+	sqlDB.Exec(t, query)
+
+	query = fmt.Sprintf("ALTER BACKUP LATEST in %s ADD NEW_KMS=%s WITH OLD_KMS=%s", userfile, key2, key1)
+	sqlDB.Exec(t, query)
+
+	ctx := context.Background()
+	store, err := execCfg.DistSQLSrv.ExternalStorageFromURI(ctx, "userfile:///a", security.RootUserName())
+	require.NoError(t, err)
+
+	files, err := getEncryptionInfoFiles(ctx, store)
+	require.NoError(t, err)
+	require.True(t, len(files) == 2)
+
+	query = fmt.Sprintf("ALTER BACKUP LATEST in %s ADD NEW_KMS=%s WITH OLD_KMS=%s", userfile, key1, key2)
+	sqlDB.Exec(t, query)
+	files, err = getEncryptionInfoFiles(ctx, store)
+	require.NoError(t, err)
+	require.True(t, len(files) == 3)
+}
+
+// TestAlterBackupRestore tests to see that an altered backup can be correctly restored using the new key,
+// and that the old key will be correctly rejected.
+func TestAlterBackupRestore(t *testing.T) {
+	defer leaktest.AfterTest(t)()
+	defer log.Scope(t).Close(t)
+
+	oldURI := getAWSEncryptionOption(t, "OLD_AWS_KMS_REGION", "OLD_AWS_KEY_ID")
+	newURI := getAWSEncryptionOption(t, "NEW_AWS_KMS_REGION", "NEW_AWS_KEY_ID")
+
+	const userfile = "'userfile:///a'"
+	const numAccounts = 1
+	t.Run(fmt.Sprintf("alter-backup-restore-with-%s", newURI), func(t *testing.T) {
+		_, sqlDB, _, cleanupFn := backupRestoreTestSetup(t, singleNode, numAccounts, InitManualReplication)
+		defer cleanupFn()
+		query := fmt.Sprintf("BACKUP TABLE bank INTO %s WITH kms = %s", userfile, oldURI)
+		sqlDB.Exec(t, query)
+
+		query = fmt.Sprintf("ALTER BACKUP LATEST in %s ADD NEW_KMS=%s WITH OLD_KMS=%s", userfile, newURI, oldURI)
+		sqlDB.Exec(t, query)
+
+		sqlDB.Exec(t, "DROP TABLE bank")
+		query = fmt.Sprintf("RESTORE TABLE bank FROM LATEST in %s WITH KMS=%s", userfile, newURI)
+		sqlDB.Exec(t, query)
+	})
+
+	t.Run(fmt.Sprintf("alter-backup-restore-with-%s", oldURI), func(t *testing.T) {
+		_, sqlDB, _, cleanupFn := backupRestoreTestSetup(t, singleNode, numAccounts, InitManualReplication)
+		defer cleanupFn()
+		query := fmt.Sprintf("BACKUP TABLE bank INTO %s WITH kms = %s", userfile, oldURI)
+		sqlDB.Exec(t, query)
+
+		query = fmt.Sprintf("ALTER BACKUP LATEST IN %s ADD NEW_KMS= %s WITH OLD_KMS = %s", userfile, newURI, oldURI)
+		sqlDB.Exec(t, query)
+
+		sqlDB.Exec(t, "DROP TABLE bank")
+		query = fmt.Sprintf("RESTORE TABLE bank FROM LATEST in %s WITH KMS = %s", userfile, oldURI)
+		sqlDB.Exec(t, query)
+	})
+}
+
+// TestAlterBackupShowBackup tests to see that show backup correctly recognizes the new encryption-info file
+// when SHOW BACKUP is called on an encrypted backup that was altered.
+func TestAlterBackupShowBackup(t *testing.T) {
```

Comment:
can I suggest a slight restructure:

- In `TestAlterBackupRestore` you can keep the create table, backup, alter backup statements
- Then create t.Run() subtests for RESTORE and SHOW that use the common backup. This way you only do the setup once and the test is more readable.

I'd also like to see tests for:
- More than one NEW_KMS
- More than one OLD_KMS (with maybe one invalid, one valid)
- And a new kms == old kms maybe? It should just work but better to test it IMO.

I'll leave it up to you to decide which of these are easy enough to add in!

Reply:
Done. Although I couldn't get the incremental backup tests to work in there since the extra files it makes was messing up some other tests. Let me know if you think this is an issue though, I can probably figure it out just wasn't sure if it was worth the effort to add in just so it could be in the same test.